static/js/stream_list.js

```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+                                    .expectOne().data('stream-id');
+
+        var top_stream = stream_data.get_sub_by_id(selected_stream_id);
+
+        if (overlays.is_active()) {
+            ui_util.change_tab_to('#home');
+        }
+        exports.clear_and_hide_search();
+        narrow.by('stream', top_stream.name,
+            {select_first_unread: true, trigger: 'sidebar enter key'});
+    }
+}
+
+function next_sibing_in_dir(elm, dir_up) {
+    if (dir_up) {
+        return elm.prev();
+    }
+
+    return elm.next();
+}
+
+function keydown_arrow_key(dir_up, all_streams_selector,
+                         scroll_container, highlighting_class) {
+    // Are there streams to cyle through?
+    if ($(all_streams_selector).length > 0) {
+
+        var current_sel = $(all_streams_selector + '.' + highlighting_class).expectOne();
+
+        var next_sibling = next_sibing_in_dir(current_sel, dir_up);
+
+        if (highlighting_class === 'highlighted_stream'
+            && next_sibling.is('hr.stream-split')) {
+            // Only for the left sidebar
+            // Skip separator
+            next_sibling = next_sibing_in_dir(next_sibling, dir_up);
+        }
 
-        var top_stream_id = $('#stream_filters li.narrow-filter').first().data('stream-id');
-        // undefined if there are no results
-        if (top_stream_id !== undefined) {
-            var top_stream = stream_data.get_sub_by_id(top_stream_id);
-            if (overlays.is_active()) {
-                ui_util.change_tab_to('#home');
+        if (!next_sibling.is('li.narrow-filter')) {
+            // At the every bottom or top
+            var all_streams = $(all_streams_selector);
+            if (dir_up) {
+                // top -> start at the bottom
+                next_sibling = all_streams.last();
+            } else {
+                // bottom -> start at the top
+                next_sibling = all_streams.first();
             }
-            exports.clear_and_hide_search();
-            narrow.by('stream', top_stream.name,
-                      {select_first_unread: true, trigger: 'sidebar enter key'});
-            e.preventDefault();
-            e.stopPropagation();
         }
+
+        // Classes must be explicitly named
+        if (highlighting_class === 'highlighted_stream') {
+            current_sel.removeClass('highlighted_stream');
+            next_sibling.addClass('highlighted_stream');
+
+        } else if (highlighting_class === 'highlighted_user') {
+            current_sel.removeClass('highlighted_user');
+            next_sibling.addClass('highlighted_user');
+        }
+
+        exports.scroll_element_into_container(next_sibling, scroll_container);
+    }
+}
+
+
+exports.keydown_filter = function (e, all_streams_selector,
+ scroll_container, highlighting_class, enter_press_function) {
+    // Function for left and right sidebar
+    // Could be placed somewhere else but ui.js is already very full
```

Comment:
Agreed.  I think we'll probably eventually want to move this and its helpers to some new file like `sidebar_ui.js` (whether that's the right name depends a bit on whether we think this could be reused for other keyboard UI situations as well).  I'd save that for a follow-up project, though, since this is already a large PR.